bot/common/threads/onboarding.py

```diff
@@ -98,56 +118,315 @@ async def handle_emoji(self, raw_reaction):
         return _handle_skip_emoji(raw_reaction, self.guild_id)
 
 
-class AddUserTwitterStep(BaseStep):
-    """Step to submit twitter name for the govrn profile"""
+class AddUserAccountStep(BaseStep):
+    """Step which prompts user for an account name. Derived classes"""
 
-    name = StepKeys.ADD_USER_TWITTER.value
+    """override the message and cache name for the response"""
 
-    def __init__(self, guild_id):
+    def __init__(self, user_id, guild_id, cache):
         super().__init__()
+        self.user_id = user_id
         self.guild_id = guild_id
+        self.cache = cache
 
     async def send(self, message, user_id):
         channel = message.channel
         sent_message = await channel.send(
-            "What twitter handle would you like to associate with this guild!"
+            "What %s would you like to associate with this guild?"
+            % self.get_account_descriptor()
         )
         return sent_message, None
 
     async def save(self, message, guild_id, user_id):
-        record_id = await find_user(message.author.id, guild_id)
-        await update_user(
-            record_id, "twitter", message.content.strip().replace("@", "")
-        )
+        metadata_key = self.get_account_storage_key()
+        account_name = self.sanitize_account_input(message)
+        await write_cache_metadata(user_id, self.cache, metadata_key, account_name)
 
-    async def handle_emoji(self, raw_reaction):
-        return _handle_skip_emoji(raw_reaction, self.guild_id)
+    @abstractmethod
+    def get_account_descriptor(self):
+        pass
 
+    @abstractmethod
+    def sanitize_account_input(self, message):
+        pass
```

Comment:
I think we should prefer composition here over inheritance. My worry is as the inheritance structure becomes deeper it will become more difficult to reason about what methods are available. At most we should aim to have one level of depth.

Reply:
done, removed the abstract method decorators and used composition to reuse this funcitonality.